ethereum/blockwatch/block_watcher.go

```diff
@@ -185,46 +200,99 @@ func (w *Watcher) GetAllRetainedBlocks() ([]*miniheader.MiniHeader, error) {
+	latestHeader, err := w.client.HeaderByNumber(nil)
 	if err != nil {
 		return err
 	}
-	if latestHeader == nil {
-		if w.startBlockDepth == rpc.LatestBlockNumber {
-			nextBlockNumber = nil // Fetch latest block
-		} else {
-			nextBlockNumber = big.NewInt(int64(w.startBlockDepth))
-		}
-	} else {
-		nextBlockNumber = big.NewInt(0).Add(latestHeader.Number, big.NewInt(1))
-	}
-	nextHeader, err := w.client.HeaderByNumber(nextBlockNumber)
+	latestBlockNumber := latestHeader.Number.Int64()
+	lastStoredHeader, err := w.stack.Peek()
 	if err != nil {
-		if err == ethereum.NotFound {
-			log.WithFields(log.Fields{
-				"blockNumber": nextBlockNumber,
-			}).Trace("block header not found")
-			return nil // Noop and wait next polling interval
-		}
 		return err
 	}
+	var lastStoredBlockNumber int64
+	if lastStoredHeader != nil {
+		lastStoredBlockNumber = lastStoredHeader.Number.Int64()
+	}
 
-	events := []*Event{}
-	events, err = w.buildCanonicalChain(nextHeader, events)
-	// Even if an error occurred, we still want to emit the events gathered since we might have
-	// popped blocks off the Stack and they won't be re-added
-	if len(events) != 0 {
-		w.blockFeed.Send(events)
+	var numBlocksToFetch int
+	// No blocks stored yet, fetch the first
+	if lastStoredHeader == nil {
+		numBlocksToFetch = 1
+	} else {
+		// Noop if already caught up or ahead of latest block returned from Ethereum node
+		if latestBlockNumber <= lastStoredBlockNumber {
+			return nil
+		}
+		numBlocksToFetch = int(latestBlockNumber - lastStoredBlockNumber)
 	}
-	if err != nil {
-		return err
+
+	if numBlocksToFetch >= constants.MaxBlocksStoredInNonArchiveNode {
+		return fmt.Errorf(
+			"cannot sync blocks more than or equal to %d blocks behind latest (%d). Your `BLOCK_POLLING_INTERVAL` is probably set too high",
+			constants.MaxBlocksStoredInNonArchiveNode,
+			numBlocksToFetch,
+		)
+	}
+
+	allEvents := []*Event{}
+	var syncErr error
+	for i := 0; i < numBlocksToFetch; i++ {
+		// Optimization: If numBlocksToFetch is 1, we already know what the nextHeader is, so avoid
+		// fetching it again. If there is more then 1 block to fetch, compute each from the last
+		// stored and fetch it
+		nextHeader := latestHeader
+		if numBlocksToFetch != 1 {
+			lastStoredHeader, err := w.stack.Peek()
+			if err != nil {
+				syncErr = err
+				break
+			}
+			nextBlockNumber := big.NewInt(0).Add(lastStoredHeader.Number, big.NewInt(1))
+			nextHeader, err = w.client.HeaderByNumber(nextBlockNumber)
+			if err != nil {
+				syncErr = err
+				break
+			}
+		}
+
+		var events []*Event
+		events, err = w.buildCanonicalChain(nextHeader, events)
+		allEvents = append(allEvents, events...)
+		if err != nil {
+			syncErr = err
+			break
+		}
+	}
+	if len(allEvents) == 0 {
+		return syncErr
+	}
+	if w.shouldRevertChanges(lastStoredHeader, allEvents) {
+		if err := w.stack.Reset(w.stackCheckpointID); err != nil {
+			return err
+		}
+	} else {
+		w.stackCheckpointID, err = w.stack.Checkpoint()
```

Comment:
I see. Found a way to get rid of the instance variable. We now checkpoint right at the beginning of the `syncChain` method, and use that checkpointID when calling `Reset()`. The only awkward part is that we also checkpoint again at the end of the method if we decide to persist the changes in the in-memory stack to the DB.

The reason why `Checkpoint()` persists the in-memory stack changes to the DB is because we are supporting a generic `Stack` interface in BlockWatcher (could be in-memory only or in-memory & DB persisted), and `Checkpoint()` is the logical place to persist to the DB, making the changes "official". 

The call to `Checkpoint()` at the beginning of `syncChain` is therefore largely to avoid the footgun issue brought up. It will always just result in incrementing the checkpointID, since the other changes made by `Checkpoint` will have already been made by the previous run of `syncChain`.